man/sat-podman.8.rst
============
 SAT-PODMAN
============

----------------------------------------------
The System Admin Toolkit Container Environment
----------------------------------------------

:Author: Hewlett Packard Enterprise Development LP.
:Copyright: Copyright 2021-2022 Hewlett Packard Enterprise Development LP.
:Manual section: 8

SYNOPSIS
========

**sat bash**

**sat** [global-opts] SUBCOMMAND [opts]

DESCRIPTION
===========

All 'sat' subcommands run a process in a container. The container can
be used to execute multiple 'sat' subcommands in a container environment
or a single 'sat' subcommand using a temporary container depending on
how the SAT command is run.

SAT subcommands may be run one of two ways:

1. First running 'sat bash', which will return a command prompt in the SAT
   container environment. At that point, 'sat' subcommands may be run.
2. Running a single 'sat' subcommand directly from the host.
   The 'sat' subcommand is run in a temporary container which
   automatically exits and is removed when the 'sat' subcommand completes.

OPTIONS
=======

For a full list of 'sat' options and arguments, access the main SAT man page
by running 'sat-man sat'. See sat\-man(8).

ENVIRONMENT
===========

Environment variables can be set to override defaults used when the container
is created. The following environment variables are available:

SAT_CONFIG_DIR
        Environment variable for the directory containing the
        SAT configuration file and authentication tokens.
        The default is $HOME/.config/sat.
        The directory is mounted in the container as $HOME/.config/sat.

SAT_KUBE_CONFIG_FILE
        Environment variable for the Kubernetes configuration file.
        The default is /etc/kubernetes/admin.conf.
        The directory is mounted in the container as $HOME/.kube/config.

SAT_LOG_DIR
        Environment variable for the directory containing the SAT log file.
        The default is /var/log/cray/sat.
        The directory is mounted in the container in the same location
        defined by SAT_LOG_DIR defaulting to /var/log/cray/sat.

FILES
=====

The 'sat' subcommands run in a container environment. When the container is
started, it automatically mounts the following host directories in the
container:

$HOME
        The home directory on the host. This is mounted in the same location
        as the host.

$PWD
        The current working directory on the host. This is mounted in
        the container as /sat/share/, and the default working directory
        of the container is set to /sat/share.

As a result, the files and directories that are located in $HOME and $PWD
are available to the subcommands that run in the container, and any files
that are created by the 'sat' subcommands in these directories will persist
on the host after the container exits. However, files in the current working
directory must be specified using relative paths to that directory, because
the current working directory is always mounted on /sat/share. Absolute paths
should be avoided, and paths that are outside of $HOME or $PWD are never
accessible to the container environment.

In addition, the following host files and directories are mounted in the container
for use by the 'sat' subcommands:

$HOME/.config/sat
        The directory containing the SAT configuration file and authentication tokens.
        The directory is created if it does not exist.
        The directory is mounted in the container as $HOME/.config/sat
        in read-write mode.
        The location of the directory on the host can be changed using the
        environment variable SAT_CONFIG_DIR.

/etc/ceph
        The directory containing ceph configuration data.
        The directory is mounted in the container as /etc/ceph
        in read-only mode.

/etc/kubernetes/admin.conf
        The file containing kubernetes configuration data.
        The file is mounted in the container as $HOME/.kube/config
        in read-only mode.
        The location of the file on the host can be changed using the
        environment variable SAT_KUBE_CONFIG_FILE.

/etc/os-release
        The file containing Operating System release information.
        The file is mounted in the container as /opt/cray/sat/etc/os-release
        in read-write mode.

/etc/pki/trust/anchors
        The directory containing CA certificates.
        The directory is mounted in the container as /usr/local/share/ca-certificates
        in read-only mode.

/opt/cray/etc
        The directory containing site information which is a local cache of the
        site information downloaded from the configured S3 bucket.
        The directory is created if it does not exist.
        The directory is mounted in the container as /opt/cray/etc
        in read-write mode.

/opt/cray/etc/release
        The directory containing product version files.
        The directory is mounted in the container as /opt/cray/etc/release
        in read-only mode.

/opt/cray/etc/sat/version
        The file containing the version of the SAT container image to use.
        When this file does not exist, the following image is used:
        registry.local/artifactory.algol60.net/sat-docker/stable/cray-sat:csm-latest

/var/log/cray/sat
        The directory containing the SAT log file.
        The directory is created if it does not exist.
        The directory is mounted in the container as /var/log/cray/sat
        in read-write mode.
        The location of the directory on the host and in the container
        can be changed using the environment variable SAT_LOG_DIR.

EXAMPLES
========

Enter the SAT container environment, then view the status of the system:

::

    # sat bash
    (CONTAINER_ID) sat-container:# sat status

Check the status of the system without entering the SAT container environment:

::

    # sat status

Show the sensor readings for the list of xnames in the xnames.txt file
located in the current working directory:

::

    # sat sensors --xname-file xnames.txt

View the full man page for 'sat':

::

    # sat-man sat

SEE ALSO
========

sat-man(8)

.. include:: _notice.rst


CHANGELOG.md
# Changelog

(C) Copyright 2020-2022 Hewlett Packard Enterprise Development LP

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2022-08-09

### Changed
- Changed the wrapper scripts to depend on a file named ``/opt/cray/etc/sat/version``
  to supply the version of the SAT container image, so that there is no longer
  a hard-coded default SAT version in the script. When this file does not
  exist, the following image is used:
  ``registry.local/artifactory.algol60.net/sat-docker/stable/cray-sat:csm-latest``
- Changed the ``SAT_REPOSITORY`` environment variable to no longer include
  part of the image name.
- Add a ``SAT_IMAGE_NAME`` variable so that the name of the image can
  be specified.

## [1.9.1] - 2022-07-05

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.17.0-20220627210641_7dd2953 to
  3.17.1-20220705195400_fd46060.

## [1.9.0] - 2022-06-27

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.16.1-20220608212829_ab4cc61 to
  3.17.0-20220627210641_7dd2953.

## [1.8.5] - 2022-06-27

### Changed

- Changed the format of the license and copyright text in all of the source
  files.

## [1.8.4] - 2022-06-15

### Fixed
- Fixed builds of stable RPMs by running publishing step outside of container,
  and add Dockerfile which sets up build dependencies, so that the build does
  not need to run as root.

## [1.8.3] - 2022-06-08

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.16.0-20220603195202_230a55e to
  3.16.1-20220608212829_ab4cc61.

## [1.8.2] - 2022-06-07

### Changed
- Changed build to publish to the ``sat-rpms`` Artifactory repository.

## [1.8.1] - 2022-06-03

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.15.1-20220418195845_1dafe87 to
   3.16.0-20220603195202_230a55e.

## [1.8.0] - 2022-05-25

### Changed
- Incremented version to 1.8.0 in preparation for SAT container image version
  3.16.
- Made changes related to open sourcing of sat-install-utility.
    - Update Jenkinsfile to use csm-shared-library.
    - Add Makefile for building RPM package.

## [1.6.8] - 2022-04-18

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.15.0-20220330153238_c5599e8 to
  3.15.1-20220418195845_1dafe87.

## [1.6.7] - 2022-04-07

### Changed
- Changed the mounting of the current working directory to mount it in the
  container as ``/sat/share/`` instead of mounting it in the container on
  the same path as on the host.
- Changed the mount of ``/etc/os-release`` to be read-only.

## [1.6.6] - 2022-03-30

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.14.0-20220224220225_70aeead to
  3.15.0-20220330153238_c5599e8.

## [1.6.5] - 2022-02-24

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.13.1-20220127225847_6b511eb to
  3.14.0-20220224220225_70aeead.

## [1.6.4] - 2022-01-27

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.13.0-20220124223033_96b5819 to
  3.13.1-20220127225847_6b511eb.

## [1.6.3] - 2022-01-24

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.12.1-20211214153306_7ab0730
  to 3.13.0-20220124223033_96b5819.

## [1.6.2] - 2021-12-14

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.12.0-20211207204308_4c03082
  to 3.12.1-20211214153306_7ab0730.

## [1.6.1] - 2021-12-07

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.11.0-20211029193657_c204587
  to 3.12.0-20211207204308_4c03082.

## [1.6.0] - 2021-11-22

### Changed
- The pager used by `sat-man` was changed from `more` to `less`.

## [1.5.2] - 2021-10-29

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.10.0-20210907172136_47bb1ee
  to 3.11.0-20211029193657_c204587.

## [1.5.1] - 2021-09-07

### Changed
- Bumped the default tag for the ``cray/cray-sat`` container image that is
  specified in the wrapper scripts from 3.9.0-20210804200331_9f1a814
  to 3.10.0-20210907172136_47bb1ee.

## [1.5.0] - 2021-08-04

### Fixed

sat-podman.sh
#!/usr/bin/env bash
#
# MIT License
#
# (C) Copyright 2021-2022 Hewlett Packard Enterprise Development LP
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
# OTHER DEALINGS IN THE SOFTWARE.
#
# Wrapper script for sat CLI

# Text surrounded by @ is replaced by a sed command in the spec file
sat_repository=${SAT_REPOSITORY:-@DEFAULT_SAT_REPOSITORY@}
sat_version_file="/opt/cray/etc/sat/version"
if [[ -f "$sat_version_file" ]]; then
    sat_image_name="${SAT_IMAGE_NAME:-@DEFAULT_SAT_IMAGE_NAME@}"
    sat_image_tag="${SAT_TAG:-$(cat "$sat_version_file")}"
else
    sat_image_name="${SAT_IMAGE_NAME:-@CSM_SAT_IMAGE_NAME@}"
    sat_image_tag="${SAT_TAG:-csm-latest}"
fi
sat_image="${SAT_IMAGE:-"${sat_repository}/${sat_image_name}:${sat_image_tag}"}"

nameserver=$(awk '{ if ($1 == "nameserver") { print $2; exit } }' /etc/resolv.conf )
sat_dns_server=${SAT_DNS_SERVER:-$nameserver}

cray_release_dir="/opt/cray/etc/release"
site_info_dir="/opt/cray/etc"
host_os_info_file="/etc/os-release"
host_os_info_target_file="/opt/cray/sat/etc/os-release"
ceph_config_dir="/etc/ceph"
cert_src_dir=${SAT_CERT_SRC_DIR:-/etc/pki/trust/anchors}
cert_target_dir=${SAT_CERT_TARGET_DIR:-/usr/local/share/ca-certificates}
kube_config_file=${SAT_KUBE_CONFIG_FILE:-/etc/kubernetes/admin.conf}
sat_config_dir=${SAT_CONFIG_DIR:-$HOME/.config/sat/}
sat_log_dir=${SAT_LOG_DIR:-/var/log/cray/sat/}

podman_cli_args="--dns $sat_dns_server"

working_dir=$(pwd -P)
if [ -d "$working_dir" ]; then
  podman_cli_args="$podman_cli_args --mount type=bind,src=$working_dir,target=/sat/share --workdir /sat/share"
fi
if [ -d "$HOME" ]; then
  podman_cli_args="$podman_cli_args --mount type=bind,src=$HOME,target=$HOME"
fi

if [ -f $host_os_info_file ]; then
  podman_cli_args="$podman_cli_args --mount type=bind,src=$host_os_info_file,target=$host_os_info_target_file,ro=true"
fi

if [ -d $cray_release_dir ]; then
  podman_cli_args="$podman_cli_args --mount type=bind,src=$cray_release_dir,target=$cray_release_dir,ro=true"
fi
if [ -d $cert_src_dir ]; then
  podman_cli_args="$podman_cli_args --mount type=bind,src=$cert_src_dir,target=$cert_target_dir,ro=true"
fi
if [ -f $kube_config_file ]; then
  podman_cli_args="$podman_cli_args --mount type=bind,src=$kube_config_file,target=$HOME/.kube/config,ro=true"
fi
if [ -d $ceph_config_dir ]; then
  podman_cli_args="$podman_cli_args --mount type=bind,src=$ceph_config_dir,target=$ceph_config_dir,ro=true"
fi

# If the site information directory does not exist and cannot be created, then give a warning.
if mkdir -p $site_info_dir; then
  podman_cli_args="$podman_cli_args --mount type=bind,src=$site_info_dir,target=$site_info_dir"
else
  echo "WARNING: Unable to create sat site information directory $site_info_dir." >&2
fi
# If configuration directory does not exist and cannot be created, then give a warning.
if mkdir -p $sat_config_dir; then
  podman_cli_args="$podman_cli_args --mount type=bind,src=$sat_config_dir,target=$HOME/.config/sat/"
else
  echo "WARNING: Unable to create sat configuration directory $sat_config_dir." \
       "No configuration file will be present." >&2
fi
# If log directory does not exist and cannot be created, then give a warning.
if mkdir -p $sat_log_dir; then
  podman_cli_args="$podman_cli_args --mount type=bind,src=$sat_log_dir,target=$sat_log_dir"
else
  echo "WARNING: Unable to create sat logging directory $sat_log_dir." \
       "No log file will be present." >&2
fi

podman_cli_args="$podman_cli_args --rm --interactive"

# allow running 'sat bash' to open a shell in sat container
if [ "$1" == "bash" ]; then
  podman run $podman_cli_args --tty $sat_image
elif [ -t 1 ]; then
  # stdout is a terminal
  podman run $podman_cli_args --tty $sat_image sat "$@"
else
  # stdout is not a terminal
  podman run $podman_cli_args $sat_image sat "$@"
fi
